### ISSUE TITLE
Particular scopes of PseudoStash'es are not clear

### ISSUE THREAD

**Issue by vrurg** [Raku/problem-solving#294]
When it comes to pseudo-packages their particular behaviors remain rather foggy in some edge cases. In some cases the scopes they cover are not what a user may expect them to be in Rakudo. For example:

```raku
my &inner;
{
    my $*FOO;
    &inner = sub {
        say OUTERS::<$*FOO>:exists;   # True
        say "Direct access: ", $*FOO; # Throws with "Dynamic variable not found"
    }
}
inner;
```

The discrepancy is caused by `OUTERS` using `nqp::getlexrel` to lookup a context's lexical scope, whereas compiler emits pure dynamic chain lookup code for `$*FOO`.

Another confusing behavior could be observed with `DYNAMIC`. I would live alone the fact that it doesn't currently work with dynamics declared on `PROCESS` and that implementing this interaction is humpered by a number of confusions caused by the fact that `$FOO` on `PROCESS` becomes `$*FOO` in Raku code. More down-to-earth uncertainty is about container's `dynamic` property which is used to determine dynamic variables. `is dynamic` trait allows one to declare a non-twigilled variable as a dynamic and it becomes accessible via dynamic chain pseudos. But this doesn't work with the compiler:

```raku
sub inner {
    say $foo; # Compile time error "Variable '$foo' is not declared"
}
sub outer {
    my $foo is dynamic = 42;
    inner;
}
outer;
```

Yet, `say DYNAMIC::.keys;` in the sub `inner` will come up with `$foo` included.

**Comment by vrurg** [comment]
# Proposal 

I would try to clarify what semantics are expected from pseudos where I consider it currently being fuzzy or wrong. In all cases I'd like it to be as close to the language semantics as possible.

Everything stated here is only applicable to `6.e`.

### `MY`

Access to all symbols declared in the current context.

### `CALLERS` and `DYNAMIC`

The only difference between the two is `DYNAMIC` including the innermost (*my*) context. Both are expected to to see all `*`-twigilled symbols except for those provided by `GLOBAL` and `PROCESS` namespaces. Non-twigilled dynamic containers are to be ignored.

**A side note** `is dynamic` trait could be deprecated as it only works through dynamic chain pseudos. To my view it is misleading. Besides, it is only roast which is using this combination. Neither CORE nor any ecosystem module are using the trait.

### `CALLER`

I'm not really sure about this one. Naming suggest it to behave like `CALLERS` but only be applied to the immediate caller frame. In this case it must only provide `*`-twigilled symbols. But I doubt it makes much sense. Up to my understanding, this namespace was introduced as the primary means of accessing `$_` in `6.c` where the topic is a dynamic. It makes less sense in `6.d`, and will make even less in `6.e` if we limit it to `*`-only symbols.

I tend to give it the same semantics which we currently achieve with `CALLER::MY`.

### `CLIENT`

Since the only way it differs from `CALLER` is how the context is picked up, I think it should follow the same semantics.

### `OUTER` and `OUTERS`

Only lexicals, no `*`-twigilled. `dynamic` container property is to be ignored.

### `LEXICAL`

It's current description in the documentation is strange. It says about _dynamic symbols_ in outer lexical scopes, but this basically has no meaning since the outers are not part of the call stack. And while the following code works, I don't think it is a good thing to have because it might result in undesirable side effects in more complicated cases:

```raku
my &s;
{
    my $*FOO = "foo!";
    &s = sub {
        say LEXICAL::<$*FOO>; # foo!
    }
}
s;
```

This semantics is also misleading because `$*FOO` is not lexically visible in the sub; see the first example in the first comment.

There are two possible approaches I consider for this namespace.

The first is rather straight forward: make it `OUTERS`+(`MY`, except for `*`-dynamics).

The second is the one I tempted to have as the accepted one. Make this namespace provide all symbols visible in a context it is applied to. This would include all lexical and dynamic ones. Apparently, if a context is a lexical outer the only dynamics visible would be those declared in the context itself. For a caller context we'd have all dynamics from that and it's callers. The only exceptions are `GLOBAL` and `PROCESS` again.

Basically, the latter semantics is what is currently implemented by `6.e` due to my wrong interpretation of the documentation. It is buggy, but `LEXICAL` is marked as both static and dynamic chain namespace.

Alternatively we can have `LEXICAL` be the first option and add a new pseudo for the second. Then a good, self-explanatory, name for the new namespace is to be invented.

## Bottom Line

By grouping into kinds of scope chains, the proposed semantics would be summed up in the following way:

- precise scope (like `MY`) includes all symbols, but only from the particular context; no callers or outers included
- dynamic chains (like `DYNAMIC`) only includes `*`-twigilled symbols, no `GLOBAL` or `PROCESS` included
- static chains (like `OUTERS`) include anything but `*`-twigilled dynamics

**Comment by vrurg** [comment]
# Proposal On "Access Via Root"

The first set of tests in _S02-names/pseudo-*.t_ is about "rooty" access, i.e. `::<$symbol>` syntax. These are currently unimplemented and not even defined.

I propose to make `::<...>` notation to be equivalent to `LEXICAL::<...>`. Presumably, this would follow the principle of the least surprise.

Explicit replacement of `::<...>` with `LEXICAL::<...>` in the test body makes it passing.

**Comment by raiph** [comment]
> The first set of tests in S02-names/pseudo-*.t is about "rooty" access, i.e. `::<$symbol>` syntax. These are currently unimplemented and not even defined.

I'm curious what you mean.

From [the doc](https://docs.raku.org/language/packages#Direct_lookup):

> The null pseudo-package is the same search list as an ordinary name search. That is, the following are all identical in meaning:
```
$foo
::{'$foo'}
::<$foo>
```
> Each of them scans the lexical scopes outward...

As far as I can tell the doc is accurate about this use of `::...` up to the point I cut it off, and the scan is the same as `LEXICAL`:
```
my $foo = 1;
{
my $baz = 3;
say ::<$foo>;      # 1
say ::<$baz>;      # 3
say $LEXICAL::foo; # 1
say $LEXICAL::baz; # 3
}
```
Filling out that last line from the doc it says:

> Each of them scans the lexical scopes outward...

> ...and then the current package scope (though the package scope is then disallowed when "strict" is in effect).

As far as I can tell that's utter tosh:
```
my $foo = 1;
{
no strict;
GLOBAL::<$bar> = 2;
my $baz = 3;
say ::<$foo>; # 1
say ::<$bar>; # Nil
say ::<$baz>; # 3
say $OUTERS::foo; # 1
say $MY::bar; # Nil
say $MY::baz; # 3
}
```

**Comment by vrurg** [comment]
I have updated the related PR. Some parts of the inital proposal are now dropped off. I would rewrite it when have time.

**Comment by raiph** [comment]
I'm working my way thru the IRC logs prior to posting summaries here. I'm going backwards, and have thus far reached 2015 in #perl6. When I reach 2005 I will then turn to the other channels. Then I also want to do something similar for the design documents. I estimate this will take me at least a week, and quite plausibly a lot longer.

> make `STATIC_CHAIN` the default mode, effectively resulting in `::` (rooty) to become a static chain

I don't like this at all. @Larry spec'd `::` as searching both lexical and package space for good reason. The fact the current implementation is deficient in this regard, and folk have been patient enough to work around that rather than complain, is not evidence that such searching is not highly desirable. But I plan to not comment further on this until I've finished my log and design doc research.

> change the semantics of `Map` `$!storage` from being a lexpad of the ctx `PseudoStash` is bound to. Now `$!storage` is a mapping from a symbol name to the lexpad where it is defined in; or it maps into an exception type if the symbol cannot be accessed via this pseudo

If this is an implementation detail that makes no semantic difference, then I can safely say I am happy the implementation is being improved! :)

If it *does* make a semantic difference, I am unable to discern what that is from this comment and would appreciate the rationale for that change being clarified before any decisions about going ahead and changing it.

Likewise with the remaining items you listed.

**Comment by vrurg** [comment]
> I don't like this at all. @larry spec'd `::` as searching both lexical and package space for good reason.

Actually, `STATIC_CHAIN` is lexical lookup. It doesn't include dynamic symbols though. Neither it includes anything else but the lexical outers. Not for now. But having it working, together with implemented binding, allowed me to de-fudge the first group of tests in _S02-names/pseudo-6e.t_. The block also has a comment:

```
    # XXX Where else should rooty access look?
    # OUR and GLOBAL are the main (mutually exclusive) choices.
```

My PR makes it possible (and easier) to add both namespaces to the lookup chain. Eventually...

> If this is an implementation detail that makes no semantic difference, then I can safely say I am happy the implementation is being improved! :)

Since all specs are passing, I guess no semantic changes are done? ;) `$!storage` is not revealed to the user space anyway. The way I use it is a compromise between memory consumption, number of additional allocations, and speed.

But talking about memory usage, I realized that there is a problem in the compiler. In case of `say MY::.keys; say MY::<$sym>` Rakudo creates the package twice, for each use. I.e. the first `MY` and the second one are two different packages. Each time a whole bunch of objects is allocated, starting with meta-objects, `PseudoStash`'es, and all the attributes. This could be fixed if pseudo-packages are installed as local symbols. But I'm not sure if it worth time spent considering that RakuAST is coming.

**Comment by vrurg** [comment]
**Everything in this note is about 6.e version of `PseudoStash`.**

Since I'm basically done with PR rakudo/rakudo#4495, I'd like to sum up changes it introduces. Eventually, I had to drop off some of the above proposed changes in semantics. The final implementation can be described based on `PseudoStash` modes.

The latest changes in the PR implement the concept of _fallbacks_. This is where a pseudo will look up for additional symbols when it's primary chain of scopes is exhausted.

# `REQUIRE_DYNAMIC`

Pseudo with this modifier can only accept dynamic symbols. I.e. either `*`-twigilled symbols, or containers with `dynamic` property set. Does not define a scope.

# `PRECISE_SCOPE`

Pseudo with mode only considers the local scope for which it is defined. Doesn't chain. By default accepts all local symbols unless combined with `REQUIRE_DYNAMIC`.

For example, `CALLER` provides all symbols in the immediate caller scope, no lexical outers or further callers included.

# `STATIC_CHAIN`

Chain of lexical outers of pseudo's scope. Uses `OUR` as its only fallback:

```raku
OUR::<$our-var> = "this is on OUR";
{
    say OUTERS::<$our-var>; # this is on OUR
}
```

Only lexicals are included. `dynamic` property of scalars is ignored making symbols like `my $dyn is dynamic` available.

# `DYNAMIC_CHAIN`

Chain of dynamic outers (callers). Includes only dynamic symbols (i.e. those `*`-twigilled), or containers with `dynamic` property set. Uses the following fallbacks:

- `$*PROMISE`, if it exists in pseudo's lexical scope and defined, and has `$!dynamic_context` set. This provides code object supplied to the method `then` with access to dynamic outers of the scope where the promise was created.
- `GLOBAL` and `PROCESS`, in the order mentioned. I.e. symbols in `GLOBAL` shadow off same symbols in `PROCESS`. Because these two namespaces contain twigil-less symbols which are then available to Raku code as twigilled (`$IN` -> `$*IN`), special measures are taken to map symbol names into twigilled notation. So, `DYNAMIC::<$*IN>` maps into `PROCESS::<$IN>`.

# Pseudo-module Modes In Raku 6.e

`::` or default - `STATIC_CHAIN`
`CALLER` - `PRECISE_SCOPE` 
`CALLERS` - `DYNAMIC_CHAIN +| REQUIRE_DYNAMIC`
`CLIENT` - `PRECISE_SCOPE`
`CORE` - `STATIC_CHAIN`
`CORE::v6<rev>` - `PRECISE_SCOPE`
`DYNAMIC` - `DYNAMIC_CHAIN +| REQUIRE_DYNAMIC`
`LEXICAL` - `STATIC_CHAIN +| DYNAMIC_CHAIN` 
`MY` - `PRECISE_SCOPE`
`OUTER` - `PRECISE_SCOPE`
`OUTERS` - `STATIC_CHAIN`
`SETTING` - `STATIC_CHAIN`
`UNIT` - `STATIC_CHAIN`, though perhaps it'd make more sense to have it as `PRECISE_SCOPE`?

*Note.* `REQUIRE_DYNAMIC` is currently can be considered useless because it's not used other than with dynamic chain pseudos, and `DYNAMIC_CHAIN` doesn't need the flag.

**Comment by vrurg** [comment]
Resolved with rakudo/rakudo#4495 and raku/roast#755.
